scripts/simplify_network.py

```diff
@@ -522,6 +523,115 @@ def consense(x):
     return clustering.network, clustering.busmap
 
 
+def drop_isolated_nodes(n, drop_e_threshold, n_periods_in_year=8760):
+    n.determine_network_topology()
+
+    generators_sum_origin = n.generators.p_nom.sum()
+    load_sum_origin = n.loads_t.p_set.sum().sum()
+
+    loads_t_ann = n_periods_in_year * n.loads_t.p_set.mean(axis=0)
+
+    # duplicated sub-networks mean that there is at least one interconnection between buses
+    i_islands = n.buses[
+        (~n.buses.duplicated(subset=["sub_network"], keep=False))
+        & (n.buses.carrier == "AC")
+    ].index
+    i_islands_drop = i_islands[loads_t_ann[i_islands] <= 1e6 * drop_e_threshold]
+
+    if i_islands_drop.empty:
+        return n
+
+    i_loads_drop = n.loads[n.loads.bus.isin(i_islands_drop)].index
+    i_generators_drop = n.generators[n.generators.bus.isin(i_islands_drop)].index
+
+    n.mremove("Bus", i_islands_drop)
+    n.mremove("Load", i_loads_drop)
+    n.mremove("Generator", i_generators_drop)
+
+    n.determine_network_topology()
+
+    load_sum_final = n.loads_t.p_set.sum().sum()
+    generators_sum_final = n.generators.p_nom.sum()
+
+    logger.info(
+        f"Dropped {len(i_islands)} buses. A resulted load discrepancy is {(100 * ((load_sum_final - load_sum_origin)/load_sum_origin)):2.1}% and {(100 * ((generators_sum_final - generators_sum_origin)/generators_sum_origin)):2.1}% for average load and generation capacity, respectivelly"
+    )
+
+    return n
+
+
+def merge_isolated_nodes(
+    n, merge_p_threshold=0.0, aggregation_strategies=dict(), n_periods_in_year=8760
+):
+    generators_sum_origin = n.generators.p_nom.sum()
+    load_sum_origin = n.loads_t.p_set.sum().sum()
+
+    n.determine_network_topology()
+
+    # duplicated sub-networks mean that there is at least one interconnection between buses
+    i_islands = n.buses[
+        (~n.buses.duplicated(subset=["sub_network"], keep=False))
+        & (n.buses.carrier == "AC")
+    ].index
+
+    # merge buses having sufficient attached load
+    loads_t_ann = n_periods_in_year * n.loads_t.p_set.mean(axis=0)
+    i_islands_large = i_islands[loads_t_ann[i_islands] >= merge_p_threshold]
+    buses_i = list(set(i_islands_large))
+
+    weight = pd.concat(
```

Comment:
Nice draft Katia:)

I saw that you kept the previous implementation using the adjacency matrix.
In practice, we don't need that.
In our case, we want to replace the calculation os busmap using our own way based on the isolated nodes.

Let's say we have 2 countries with 5 nodes total of which a1, a2 and b1 are isolated and to be merged
A1
A2
A3 (not isolated)
B1
B2 (not isolated)

The key is to create a busmap as follows:
a1->"newa1"
a2->"newa1"
a3->"newa2"
b1->"newb1"
b2->"newb2'

Please note that we cannot mix buses belonging to different countries.
The above does not require adjacency matrix calculation.

As a comment, in this testing, you have now the same setting for both the drop and merge options.
If the drop option is consolidated you can deactivate it while testing the merge.
Once also the latter works that the final version may be enabled

Reply:
> Nice draft Katia:)
> 
> I saw that you kept the previous implementation using the adjacency matrix. In practice, we don't need that. In our case, we want to replace the calculation os busmap using our own way based on the isolated nodes.
> 
> Let's say we have 2 countries with 5 nodes total of which a1, a2 and b1 are isolated and to be merged A1 A2 A3 (not isolated) B1 B2 (not isolated)
> 
> The key is to create a busmap as follows: a1->"newa1" a2->"newa1" a3->"newa2" b1->"newb1" b2->"newb2'
> 
> Please note that we cannot mix buses belonging to different countries. The above does not require adjacency matrix calculation.
> 
> As a comment, in this testing, you have now the same setting for both the drop and merge options. If the drop option is consolidated you can deactivate it while testing the merge. Once also the latter works that the final version may be enabled

Thanks a lot for explanations Davide!
Now it's much clearer with busmaps :)

As for replacement for adjacency matrix, my first thought was to build a matrix of the distances for each point and take the minimal values. But probably using KD-trees from scipy would be a more elegant approach...

Absolutely regarding the thresholds. Thanks for the hint!

Reply:
So, the use for the adjacency matrix in that function is justified by the fact that that function aimed to cluster substations to their closest substation, thus it needed the adjacency matrix.

In our case, we really don't need that.
By knowing for each country the buses that are isolated, you know that those buses will have to be mapped into the same bus.
I don't see where you would need to use the KD trees